### PR TITLE
Exposing optional replaceMissingValueWith in lookup function and macros

### PR DESCRIPTION
**Pull request by pranavbhole**
### Description

Adding the optional 3rd argument replaceMissingValueWith to lookup extractor function that replaces the missing value from given lookups. Default value for replaceMissingValueWith is null. It is already supported in native layer. Just exposing this value to populated in SQL layer as well. On same line, updated the lookup macros to support replaceMissingValueWith. 

#### Release note
Lookup function also accepts the 3rd argument called '$replaceMissingValueWith' as constant string, if you value is missing given lookups for queried key then lookup function return result value from replaceMissingValueWith eg:

LOOKUP(store, 'store_to_country', 'NA')

If value is missing from store_to_country lookup for given key 'store' then it will return 'NA'.

This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [x] added documentation for new or modified features or behaviors.
- [x] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [x] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

File: docs/querying/lookups.md [hunk 1]
```diff
@@ -62,6 +62,12 @@ SELECT
 FROM sales
 GROUP BY 1
 ```
+Lookup function also accepts the 3rd argument called `$replaceMissingValueWith` as constant string, if you value is missing given lookups for queried key then lookup function return result value from `replaceMissingValueWith`
```
Comment (soumyava) [inline docs/querying/lookups.md:65]: nit. The lookup function also accepts the 3rd argument called `$replaceMissingValueWith` as a constant string. If your value is missing a lookup for the queried key, the lookup function returns the result value from `replaceMissingValueWith`

File: processing/src/main/java/org/apache/druid/query/expression/LookupExprMacro.java [hunk 2]
```diff
@@ -116,4 +126,15 @@ public void decorateCacheKeyBuilder(CacheKeyBuilder builder)
 
     return new LookupExpr(arg);
   }
+
+  private String getReplaceMissingValueWith(final List<Expr> args)
+  {
+    if (args.size() > 2) {
+      final Expr missingValExpr = args.get(2);
+      if (missingValExpr.isLiteral()) {
+        return missingValExpr.getLiteralValue().toString();
```
Comment (soumyava) [inline processing/src/main/java/org/apache/druid/query/expression/LookupExprMacro.java:135]: Q. when used in sqlCompatibleMode if there is a value like empty string should it be converted to null ?

Comment (pranavbhole) [reply processing/src/main/java/org/apache/druid/query/expression/LookupExprMacro.java:135]: Nope, regardless of the sqlCompatibleMode, replaceMissingValueWith is literal value passed by user and feature expect to replace it as it is.

File: docs/querying/lookups.md (hunk 1 above)
Comment (pranavbhole) [reply docs/querying/lookups.md:65]: fixed

File: processing/src/main/java/org/apache/druid/query/expression/LookupExprMacro.java [hunk 3]
```diff
@@ -54,10 +54,11 @@ public String name()
   @Override
   public Expr apply(final List<Expr> args)
   {
-    validationHelperCheckArgumentCount(args, 2);
+    validationHelperCheckMinArgumentCount(args, 2);
```
Comment (clintropolis) [inline processing/src/main/java/org/apache/druid/query/expression/LookupExprMacro.java:57]: this should use maybe `validationHelperCheckArgumentRange` since it requires 2 or 3 arguments exactly

File: processing/src/main/java/org/apache/druid/query/expression/LookupExprMacro.java [hunk 4]
```diff
@@ -116,4 +126,15 @@ public void decorateCacheKeyBuilder(CacheKeyBuilder builder)
...
+
+  private String getReplaceMissingValueWith(final List<Expr> args)
+  {
+    if (args.size() > 2) {
+      final Expr missingValExpr = args.get(2);
+      if (missingValExpr.isLiteral() && missingValExpr.getLiteralValue() != null) {
+        return missingValExpr.getLiteralValue().toString();
+      }
+    }
+    return null;
```
Comment (clintropolis) [inline processing/src/main/java/org/apache/druid/query/expression/LookupExprMacro.java:138]: I think if you are not going to handle a dynamic argument for the 'missing value' parameter, this should be an exception if the argument is not a literal. There is `validationHelperCheckArgIsLiteral` that can be called if we only want to support this argument being a constant.

Silently returning null seems incorrect

File: docs/querying/lookups.md [hunk 5]
```diff
@@ -62,6 +62,12 @@ SELECT
 FROM sales
 GROUP BY 1
 ```
+The lookup function also accepts the 3rd argument called `replaceMissingValueWith` as a constant string. If your value is missing a lookup for the queried key, the lookup function returns the result value from `replaceMissingValueWith`
+For example:
+```
+LOOKUP(store, 'store_to_country', 'NA')
+```
+If value is missing from `store_to_country` lookup for given key 'store' then it will return `NA`.
```
Comment (clintropolis) [inline docs/querying/lookups.md:70]: you should probably indicate that this optional 3rd argument must be a constant.

`math-expr.md`, `sql-functions.md`, and `sql-scalar.md` should also be updated since they all document this function signature

Comment (pranavbhole) [reply docs/querying/lookups.md:70]: fixed

File: processing/src/main/java/org/apache/druid/query/expression/LookupExprMacro.java (hunk 4 above)
Comment (pranavbhole) [reply processing/src/main/java/org/apache/druid/query/expression/LookupExprMacro.java:137]: added validationHelperCheckArgIsLiteral check

File: processing/src/main/java/org/apache/druid/query/expression/LookupExprMacro.java (hunk 3 above)
Comment (pranavbhole) [reply processing/src/main/java/org/apache/druid/query/expression/LookupExprMacro.java]: done

**Review by soumyava — APPROVED** [review]
LGTM

File: processing/src/test/java/org/apache/druid/query/expression/LookupExprMacroTest.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+            return Optional.empty();
+          }
+        })
+    );
+  }
+
+  @Test
+  public void testTooFewArgs()
+  {
+    expectException(IllegalArgumentException.class, "Function[lookup] requires 2 to 3 arguments");
```
Comment (clintropolis) [inline processing/src/test/java/org/apache/druid/query/expression/LookupExprMacroTest.java:63]: i suppose if the range is 1 we could change this messaging to be like `2 or 3` but we don't have to worry about that in this PR

File: processing/src/main/java/org/apache/druid/query/expression/LookupExprMacro.java [hunk 7]
```diff
@@ -116,4 +126,14 @@ public void decorateCacheKeyBuilder(CacheKeyBuilder builder)
 
     return new LookupExpr(arg);
   }
+
+  private String getReplaceMissingValueWith(final List<Expr> args)
+  {
+    if (args.size() > 2) {
+      final Expr missingValExpr = args.get(2);
+      validationHelperCheckArgIsLiteral(missingValExpr, "third argument");
+      return String.valueOf(missingValExpr.getLiteralValue());
```
Comment (clintropolis) [inline processing/src/main/java/org/apache/druid/query/expression/LookupExprMacro.java:135]: hmm, sorry I didn't notice this earlier, I think this should use `Evals.asString` so that if the value is `null` it doesn't become `"null"` as in https://github.com/apache/druid/pull/14956/files#diff-989c5e92f71c63c48629a62588a78bec198184ff896791058509176dffe91338R54

File: sql/src/main/java/org/apache/druid/sql/calcite/expression/builtin/QueryLookupOperatorConversion.java [hunk 8]
```diff
@@ -91,4 +103,18 @@ public DruidExpression toDruidExpression(
...
+  private String getReplaceMissingValueWith(
+      final List<DruidExpression> inputExpressions,
+      final PlannerContext plannerContext
+  )
+  {
+    if (inputExpressions.size() > 2) {
+      final Expr missingValExpr = plannerContext.parseExpression(inputExpressions.get(2).getExpression());
+      if (missingValExpr.isLiteral()) {
+        return missingValExpr.getLiteralValue().toString();
+      }
```
Comment (clintropolis) [inline sql/src/main/java/org/apache/druid/sql/calcite/expression/builtin/QueryLookupOperatorConversion.java:116]: this should maybe have a similar validation to the native layer to make sure that the argument doesn't silently become null in the event that the expression is null

File: sql/src/main/java/org/apache/druid/sql/calcite/expression/builtin/QueryLookupOperatorConversion.java [hunk 9]
```diff
@@ -91,4 +103,18 @@ public DruidExpression toDruidExpression(
...
+
+  private String getReplaceMissingValueWith(
+      final List<DruidExpression> inputExpressions,
+      final PlannerContext plannerContext
+  )
+  {
+    if (inputExpressions.size() > 2) {
+      final Expr missingValExpr = plannerContext.parseExpression(inputExpressions.get(2).getExpression());
+      if (missingValExpr.isLiteral()) {
+        return missingValExpr.getLiteralValue().toString();
```
Comment (clintropolis) [inline sql/src/main/java/org/apache/druid/sql/calcite/expression/builtin/QueryLookupOperatorConversion.java:115]: can this argument be null?

File: sql/src/main/java/org/apache/druid/sql/calcite/expression/builtin/QueryLookupOperatorConversion.java [hunk 10]
```diff
@@ -30,16 +30,27 @@
...
       .operatorBuilder("LOOKUP")
-      .operandTypes(SqlTypeFamily.CHARACTER, SqlTypeFamily.CHARACTER)
+      .operandTypeChecker(
+          BasicOperandTypeChecker.builder()
+                                 .operandTypes(
+                                     SqlTypeFamily.CHARACTER,
+                                     SqlTypeFamily.CHARACTER,
+                                     SqlTypeFamily.CHARACTER
+                                 )
+                                 .requiredOperandCount(2)
```
Comment (gianm) [inline sql/src/main/java/org/apache/druid/sql/calcite/expression/builtin/QueryLookupOperatorConversion.java:52]: Since the lookup expr only supports a literal "missing value replacement", this should include `.literalOperands(2)` (meaning operand 2 must be a literal).

File: docs/querying/sql-scalar.md [hunk 11]
```diff
@@ -98,7 +98,7 @@ String functions accept strings, and return a type appropriate to the function.
 |`CHAR_LENGTH(expr)`|Alias for `LENGTH`.|
 |`CHARACTER_LENGTH(expr)`|Alias for `LENGTH`.|
 |`STRLEN(expr)`|Alias for `LENGTH`.|
-|`LOOKUP(expr, lookupName)`|Look up `expr` in a registered [query-time lookup table](lookups.md). Note that lookups can also be queried directly using the [`lookup` schema](sql.md#from).|
+|`LOOKUP(expr, lookupName, [replaceMissingValueWith])`|Look up `expr` in a registered [query-time lookup table](lookups.md). Note that lookups can also be queried directly using the [`lookup` schema](sql.md#from).|
```
Comment (gianm) [inline docs/querying/sql-scalar.md:101]: Update the text to say what `replaceMissingValueWith` does, and to point out that it must be a constant.

File: processing/src/main/java/org/apache/druid/query/expression/LookupExprMacro.java [hunk 12]
```diff
@@ -104,6 +105,15 @@ public ExpressionType getOutputType(InputBindingInspector inspector)
       @Override
       public String stringify()
       {
+        if (replaceMissingValueWith != null) {
+          return StringUtils.format(
+              "%s(%s, %s, '%s')",
+              FN_NAME,
+              arg.stringify(),
+              lookupExpr.stringify(),
+              replaceMissingValueWith
```
Comment (gianm) [inline processing/src/main/java/org/apache/druid/query/expression/LookupExprMacro.java:114]: This doesn't handle escaping properly. `replaceMissingValueWith` might contain `'` itself, or newlines, etc.

The simplest way to fix this is to keep the expr around and `stringify()` it here, like we do with `lookupExpr` (which is also required to be a literal).

File: processing/src/main/java/org/apache/druid/query/expression/LookupExprMacro.java (hunk 7 above)
Comment (pranavbhole) [reply processing/src/main/java/org/apache/druid/query/expression/LookupExprMacro.java]: fixed, retained the original expression here. also checking literal in macro and added test to make sure that null literalValue is passed correctly same as native layer.

File: sql/src/main/java/org/apache/druid/sql/calcite/expression/builtin/QueryLookupOperatorConversion.java (hunk 8 above)
Comment (pranavbhole) [reply sql/src/main/java/org/apache/druid/sql/calcite/expression/builtin/QueryLookupOperatorConversion.java]: Added a test in CalciteQueryTest to make sure that null is wired up correctly to native extraction fn.  
modified the returned value to`(String) missingValExpr.getLiteralValue()`

File: sql/src/main/java/org/apache/druid/sql/calcite/expression/builtin/QueryLookupOperatorConversion.java (hunk 9 above)
Comment (pranavbhole) [reply sql/src/main/java/org/apache/druid/sql/calcite/expression/builtin/QueryLookupOperatorConversion.java]: yes it can be null, newly added test validates the correct behavior. 
Thank you

File: sql/src/main/java/org/apache/druid/sql/calcite/expression/builtin/QueryLookupOperatorConversion.java (hunk 10 above)
Comment (pranavbhole) [reply sql/src/main/java/org/apache/druid/sql/calcite/expression/builtin/QueryLookupOperatorConversion.java:52]: fixed

File: docs/querying/sql-scalar.md (hunk 11 above)
Comment (pranavbhole) [reply docs/querying/sql-scalar.md]: fixed

File: processing/src/main/java/org/apache/druid/query/expression/LookupExprMacro.java (hunk 12 above)
Comment (pranavbhole) [reply processing/src/main/java/org/apache/druid/query/expression/LookupExprMacro.java]: fixed

File: processing/src/test/java/org/apache/druid/query/expression/LookupExprMacroTest.java (hunk 6 above)
Comment (pranavbhole) [reply processing/src/test/java/org/apache/druid/query/expression/LookupExprMacroTest.java:63]: I think Range 1, will also throw the same message.

File: docs/querying/math-expr.md [hunk 13]
```diff
@@ -82,7 +82,7 @@ The following built-in functions are available.
 |concat|concat(expr, expr...) concatenate a list of strings|
 |format|format(pattern[, args...]) returns a string formatted in the manner of Java's [String.format](https://docs.oracle.com/javase/8/docs/api/java/lang/String.html#format-java.lang.String-java.lang.Object...-).|
 |like|like(expr, pattern[, escape]) is equivalent to SQL `expr LIKE pattern`|
-|lookup|lookup(expr, lookup-name) looks up expr in a registered [query-time lookup](../querying/lookups.md)|
+|lookup|lookup(expr, lookup-name[,replaceMissingValueWith]) looks up expr in a registered [query-time lookup](../querying/lookups.md)|
```
Comment (clintropolis) [inline docs/querying/math-expr.md:85]: nit: should probably specify `replaceMissingValueWith` must be a constant

File: docs/querying/sql-functions.md [hunk 14]
```diff
@@ -885,7 +885,7 @@ Calculates the base-10 of the numeric expression.
 
 ## LOOKUP
 
-`LOOKUP(<CHARACTER>, <CHARACTER>)`
+`LOOKUP(<CHARACTER>, <CHARACTER>, [<CHARACTER>])`
```
Comment (clintropolis) [inline docs/querying/sql-functions.md:888]: nit: 
```suggestion
`LOOKUP(<CHARACTER>, <CHARACTER>[, <CHARACTER>])`
```

although looking at this page, this isn't actually consistent across all functions... i think this form makes more sense because the comma wouldn't be there unless the argument is there, but :shrug:

File: docs/querying/sql-scalar.md [hunk 15]
```diff
@@ -98,7 +98,7 @@ String functions accept strings, and return a type appropriate to the function.
 |`CHAR_LENGTH(expr)`|Alias for `LENGTH`.|
 |`CHARACTER_LENGTH(expr)`|Alias for `LENGTH`.|
 |`STRLEN(expr)`|Alias for `LENGTH`.|
-|`LOOKUP(expr, lookupName)`|Look up `expr` in a registered [query-time lookup table](lookups.md). Note that lookups can also be queried directly using the [`lookup` schema](sql.md#from).|
+|`LOOKUP(expr, lookupName, [replaceMissingValueWith])`|Look up `expr` in a registered [query-time lookup table](lookups.md). Note that lookups can also be queried directly using the [`lookup` schema](sql.md#from). Optional constant replaceMissingValueWith can be passed as 3rd argument to be returned when value is missing from lookup.|
```
Comment (clintropolis) [inline docs/querying/sql-scalar.md:101]: nit:

```suggestion
|`LOOKUP(expr, lookupName[, replaceMissingValueWith])`|Look up `expr` in a registered [query-time lookup table](lookups.md). Note that lookups can also be queried directly using the [`lookup` schema](sql.md#from). Optional constant `replaceMissingValueWith` can be passed as the 3rd argument to be returned when value is missing from lookup.|
```

File: processing/src/main/java/org/apache/druid/query/expression/LookupExprMacro.java [hunk 16]
```diff
@@ -69,7 +70,9 @@ public Expr apply(final List<Expr> args)
         lookupExtractorFactoryContainerProvider,
         lookupName,
         false,
-        null,
+        replaceMissingValueWith != null && replaceMissingValueWith.isLiteral()
+        ? (String) replaceMissingValueWith.getLiteralValue()
```
Comment (clintropolis) [inline processing/src/main/java/org/apache/druid/query/expression/LookupExprMacro.java:74]: I suppose since other types can also be literals it might be safer to use `Evals.asString` instead of casting

File: sql/src/main/java/org/apache/druid/sql/calcite/expression/builtin/QueryLookupOperatorConversion.java [hunk 17]
```diff
@@ -91,4 +104,18 @@ public DruidExpression toDruidExpression(
...
+
+  private String getReplaceMissingValueWith(
+      final List<DruidExpression> inputExpressions,
+      final PlannerContext plannerContext
+  )
+  {
+    if (inputExpressions.size() > 2) {
+      final Expr missingValExpr = plannerContext.parseExpression(inputExpressions.get(2).getExpression());
+      if (missingValExpr.isLiteral()) {
+        return (String) missingValExpr.getLiteralValue();
```
Comment (clintropolis) [inline sql/src/main/java/org/apache/druid/sql/calcite/expression/builtin/QueryLookupOperatorConversion.java:116]: otoh, this one is probably fine as is and doesn't need to use `Evals.asString` like the native layer should since the operand type checker specifies the 3rd argument as character type, though it would also probably be harmless

File: sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java [hunk 18]
```diff
@@ -8967,6 +8967,65 @@ public void testFilterAndGroupByLookup()
...
+                        .setAggregatorSpecs(
+                            aggregators(
+                                new CountAggregatorFactory("a0")
+                            )
+                        )
+                        .setContext(QUERY_CONTEXT_DEFAULT)
+                        .build()
+        ),
+        ImmutableList.of(
+            new Object[]{"Missing_Value", NullHandling.sqlCompatible() ? null : "", 5L},
```
Comment (clintropolis) [inline sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java:9024]: nit: can use `NullHandling.defaultStringValue()`

**Review by clintropolis — APPROVED** [review]

File: sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java (hunk 18 above)
Comment (pranavbhole) [reply sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java]: fixed

File: sql/src/main/java/org/apache/druid/sql/calcite/expression/builtin/QueryLookupOperatorConversion.java (hunk 17 above)
Comment (pranavbhole) [reply sql/src/main/java/org/apache/druid/sql/calcite/expression/builtin/QueryLookupOperatorConversion.java]: fixed

File: processing/src/main/java/org/apache/druid/query/expression/LookupExprMacro.java (hunk 16 above)
Comment (pranavbhole) [reply processing/src/main/java/org/apache/druid/query/expression/LookupExprMacro.java]: fixed

File: docs/querying/math-expr.md (hunk 13 above)
Comment (pranavbhole) [reply docs/querying/math-expr.md]: fixed
